internal/pkg/cli/env_delete.go

```diff
@@ -242,13 +243,30 @@ func (o *deleteEnvOpts) ensureRolesAreRetained() error {
 		}
 		return fmt.Errorf("get template body for environment %s in application %s: %v", o.name, o.appName, err)
 	}
-	// Check if the roles are already being retained.
-	retainsExecRole := strings.Contains(body, `
-  CloudformationExecutionRole:
-    DeletionPolicy: Retain`)
-	retainsManagerRole := strings.Contains(body, `
-  EnvironmentManagerRole:
-    DeletionPolicy: Retain`)
+
+	// Check if the execution role and the manager role are retained by the stack.
+	const retainPolicy = "Retain"
+	roles := struct {
+		ExecRole    yaml.Node `yaml:"CloudformationExecutionRole"`
```

Comment:
Does this work for non-top-level yaml keys? I thought we had to unmarshal a `Resources` struct which contains the `CloudformationExecutionRole` and `EnvironmentManagerRole` keys.

Reply:
nope not needed! it can directly unmarshal to a sub-field. In fact there is a test for it :) https://github.com/aws/copilot-cli/pull/1799/files#diff-3afdf3932976f8d5c7f543d8af515690d276e9d965078b0bc3dd271404929a37R235

Reply:
welp you were right 🤦 , I thought that testcase was catching it and it wasn't haha because I expected those roles to not be retained >_> I'll send a fix with a `Resources` struct like you suggested